src/Rings/orderings.jl

```diff
@@ -1717,33 +1717,91 @@ end
 
 # Module orderings (not module Orderings)
 
+# For ordering the generators (I think)
 mutable struct ModOrdering{T} <: AbsModOrdering
-   gens::T
-   ord::Symbol
-   function ModOrdering(u::T, s::Symbol) where {T <: AbstractVector{Int}}
-     r = new{T}()
-     r.gens = u
-     r.ord = s
-     return r
-   end
+  gens::T
+  ord::Symbol
+  function ModOrdering(u::T, s::Symbol) where {T <: AbstractVector{Int}}
+    r = new{T}()
+    r.gens = u
+    r.ord = s
+    return r
+  end
 end
 
 mutable struct ModuleOrdering{S}
-   M::S
-   o::AbsModOrdering # must allow gen*mon or mon*gen product ordering
+  M::S
+  o::AbsModOrdering # must allow gen*mon or mon*gen product ordering
+
+  canonical_matrix::ZZMatrix
+
+  function ModuleOrdering(M::S, o::AbsModOrdering) where {S}
+    return new{S}(M, o)
+  end
 end
 
 base_ring(a::ModuleOrdering) = a.M
 
 mutable struct ModProdOrdering <: AbsModOrdering
-   a::AbsOrdering
-   b::AbsOrdering
+  a::AbsOrdering
+  b::AbsOrdering
 end
 
 Base.:*(a::AbsGenOrdering, b::AbsModOrdering) = ModProdOrdering(a, b)
 
 Base.:*(a::AbsModOrdering, b::AbsGenOrdering) = ModProdOrdering(a, b)
 
+# For equality checking and hashing
+# we produce a matrix representation by embedding the underlying free module (and its ordering) into a polynomial ring
+# then we build the matrix in this ring
+
+function _embedded_ring_ordering(o::ModuleOrdering)
+  return _embedded_ring_ordering(o.o)
+end
+
+function _embedded_ring_ordering(o::ModOrdering)
+  return SymbOrdering(o.ord, o.gens)
+end
+
+function _embedded_ring_ordering(o::AbsGenOrdering)
+  return deepcopy(o)
+end
+
+function _embedded_ring_ordering(o::ModProdOrdering)
+  ea = _embedded_ring_ordering(o.a)
+  shift = maximum(ea.vars)
+  eb = _embedded_ring_ordering(o.b)
+  eb.vars .+= shift 
+  return ea*eb
+end
+
+function _canonical_matrix_intern(o::ModuleOrdering)
+  nvrs = ngens(o.M) + ngens(base_ring(o.M))
+  eo = _embedded_ring_ordering(o)
+  return canonical_matrix(nvrs, eo)
+end
+
+function canonical_matrix(o::ModuleOrdering)
+  if !isdefined(o, :canonical_matrix)
+    if isone(ngens(o.M))
+      o.canonical_matrix = canonical_matrix(induced_ring_ordering(o))
+    else
+      o.canonical_matrix = _canonical_matrix_intern(o)
+    end
+  end
+  return o.canonical_matrix
+end
+
+function Base.:(==)(o1::ModuleOrdering, o2::ModuleOrdering)
+  return o1.M == o2.M && canonical_matrix(o1) == canonical_matrix(o2)
+end
+
+function Base.hash(o::ModuleOrdering, h::UInt)
+  h1 = hash(o.M, h)
+  h2 = hash(canonical_matrix(o), h)
+  return hash(h2, h)
```

Comment:
```suggestion
  h = hash(o.M, h)
  h = hash(canonical_matrix(o), h)
  return h
```
`h1` is not used anywhere. I think you wanted to do something like this

Reply:
that makes more sense, yes. Fixed.